### PR TITLE
feat: apply with multiprocessing

### PR DESCRIPTION
**Pull request by maartenbreddels**
POC from #1074

### CODE REVIEW

File: docs/source/tutorial.ipynb [hunk 1]
```diff
@@ -3687,6 +3687,294 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# The escape hatch: apply\n",
+    "In case a calculation cannot be expressed in a Vaex expression, the last resort method is to use apply. This can be useful if the function you want to apply is written in pure Python, and difficult or impossible to vectorize.\n",
+    "\n",
+    "We think apply should only be used as a last resort, because it needs to use multiprocessing (which spawns new processes) to avoid the Python Global Interpreter Lock (GIL) to make use of multiple cores. This comes at a cost of having to transfer the data between the main and child processes.\n",
+    "\n",
+    "In case you really want, or need to use it, here is how you use it:"
```
Comment (JovanVeljanoski) [inline docs/source/tutorial.ipynb:3700]:
```suggestion
    "Here is an example which uses the `apply` method:"
```

File: docs/source/tutorial.ipynb [hunk 2]
```diff
@@ -3687,6 +3687,294 @@
...
+    }
+   ],
+   "source": [
+    "import vaex\n",
+    "\n",
+    "def slow_is_prime(x):\n",
+    "    return x > 1 and all((x % i) != 0 for i in range(2, x))\n",
+    "\n",
+    "df = vaex.from_arrays(x=vaex.vrange(0, 100_000, dtype='i4'))\n",
+    "# you need to explicitly tell which arguments you need\n",
```
Comment (JovanVeljanoski) [inline docs/source/tutorial.ipynb:3760]:
```suggestion
    "# you need to explicitly specify which arguments you need\n",
```

File: docs/source/tutorial.ipynb [hunk 3]
```diff
@@ -3687,6 +3687,294 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are 9592 prime number between 0 and 100000\n"
+     ]
+    }
+   ],
+   "source": [
+    "prime_count = df.is_prime.sum()\n",
+    "print(f'There are {prime_count} prime number between 0 and {len(df)}')"
```
Comment (JovanVeljanoski) [inline docs/source/tutorial.ipynb:3785]:
```suggestion
    "print(f'There are {prime_count} prime numbers between 0 and {len(df)}')"
```

File: docs/source/tutorial.ipynb [hunk 4]
```diff
@@ -3687,6 +3687,294 @@
...
+    "df['is_prime'] = df.apply(slow_is_prime, arguments=[df.x])\n",
+    "# but this form only works for a single argument\n",
+    "df['is_prime'] = df.x.apply(slow_is_prime)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## When not to use apply\n",
```
Comment (JovanVeljanoski) [inline docs/source/tutorial.ipynb:3809]:
```suggestion
    "## When not to use `apply`\n",
```

File: docs/source/tutorial.ipynb [hunk 5]
```diff
@@ -3687,6 +3687,294 @@
...
+    "When your function can easily be vectorized, you should not use apply. When you use Vaex' expression system, we know what you do, we see the expression, and can manipulate it. An apply function is like a black box, we cannot do anything with it, like JIT-ting for instance."
```
Comment (JovanVeljanoski) [inline docs/source/tutorial.ipynb:3810]:
```suggestion
    "You should not use `apply` when your function can be vectorized. When you use Vaex' expression system, we know what you do, we see the expression, and can manipulate it in order to achieve optimal performance. An `apply` function is like a black box, we cannot do anything with it, like JIT-ting for instance."
```

File: docs/source/tutorial.ipynb [hunk 6]
```diff
@@ -3687,6 +3687,294 @@
...
+   "execution_count": 11,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2020-11-27T18:59:37.287763Z",
+     "start_time": "2020-11-27T18:59:37.286036Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# bad idea: same performance as just dy['y'], but we lose the information about what you have done\n",
```
Comment (JovanVeljanoski) [inline docs/source/tutorial.ipynb:3952]:
```suggestion
    "# bad idea: same performance as just dy['y'], but we lose the information about what was done\n",
```

File: docs/source/tutorial.ipynb [hunk 7]
```diff
@@ -3687,6 +3687,294 @@
...
+   "execution_count": 7,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2020-11-27T18:59:27.623180Z",
+     "start_time": "2020-11-27T18:59:27.621623Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# will transfer the data to child processes, and execute the ** in Python it for each element\n",
```
Comment (JovanVeljanoski) [inline docs/source/tutorial.ipynb:3876]:
```suggestion
    "# will transfer the data to child processes, and execute the **  operation in Python for each element\n",
```

File: packages/vaex-core/vaex/dataframe.py [hunk 8]
```diff
@@ -384,6 +384,7 @@ def apply(self, f, arguments=None, dtype=None, delay=False, vectorize=False):
 
         :param f: The function to be applied
         :param arguments: List of arguments to be passed on to the function f.
+        :param bool multiprocessing: Use multiple processes to avoid the GIL (Global interpreter lock)
```
Comment (JovanVeljanoski) [inline packages/vaex-core/vaex/dataframe.py:387]: Should we take this opportunity to document the rest of the kwargs here?

File: packages/vaex-core/vaex/expression.py [hunk 9]
```diff
@@ -856,7 +856,7 @@ def apply(self, f):
         :param f: A function to be applied on the Expression values
```
Comment (JovanVeljanoski) [inline packages/vaex-core/vaex/expression.py:856]: Should we document all kwargs here?

File: docs/source/tutorial.ipynb (hunk 4 above)
Comment (maartenbreddels) [reply docs/source/tutorial.ipynb:3809]: this never format correctly with nbsphinx

File: packages/vaex-core/vaex/dataframe.py (hunk 8 above)
Comment (maartenbreddels) [reply packages/vaex-core/vaex/dataframe.py:387]: yeah, vectorize at least

File: packages/vaex-core/vaex/expression.py (hunk 9 above)
Comment (maartenbreddels) [reply packages/vaex-core/vaex/expression.py:856]: yeah

**Pull request merged**